README.md
mkdong
======

A Python module for printing dongs.  
_This module borrows heavily from https://github.com/jathanism/mkdong.git_  


__installation:__  
```bash
$ git clone https://github.com/cfralick/mkdong.git  
$ cd mkdong  
$ python setup.py install  
```    

__usage:__
```bash
$ python -m mkdong 5       
```

```python
>>> import mkdong     
>>> mkdong.dong(5)    
'( )/( )=====D'    
```

__help:__
```bash
$ python -m mkdong -h
usage: mkdong [-h] [-w] [-c] [-v] l

Prints a dong.

positional arguments:
  l              the desired dong length

optional arguments:
  -h, --help     show this help message and exit
  -w, --wide     make a wide, thick dong
  -c, --climax   makes the dong climax
  -v, --version  print version and exit

mkdong 0.0.1
```


CHANGELOG.rst
#########
Changelog
#########

7.0 (2016-03-02)
================

+ Total refactor and added docstrings to everything!

6.1 (2015-05-29)
================

+ Now with unit tests!

6.0 (2014-03-01)
================

+ Added support for longdongs.

5.0 (2014-02-24)
================

+ Fixed issue with dongs being too big.

Older Versions
==============

+ Who cares?


tests/test_mkdong.py
import unittest
from mkdong.core import mkdong, MAXLEN
from argparse import ArgumentError


class MkdongTestCase(unittest.TestCase):

    def setUp(self):
        self.default_dong = mkdong(1, False)
        self.maxlen = MAXLEN
        self.dong_too_long = MAXLEN + 1

    def tearDown(self):
        self.default_dong = None
        self.maxlen = 0
        self.dong_too_long = None

    def test_dong_is_dong(self):
        dong = '( )/( )=D'
        self.assertEqual(mkdong(1, False), dong)

    def test_wide_dong_is_wide(self):
        dong = '( )/( )/D'
        self.assertEqual(mkdong(1, True), dong)

    def test_microdong_is_still_dong(self):
        micro = '( )/( )D'
        self.assertEqual(mkdong(0, False), micro)

    def test_wide_thin_microdongs_same(self):
        self.assertEqual(mkdong(0, False), mkdong(0, True))

    def test_climaxing_dong_shows_load(self):
        self.assertEqual(mkdong(1, False, True), self.default_dong + ' ~~~~')

    def test_non_int_dong_length_raises_type_error(self):
        with self.assertRaises(TypeError):
            mkdong('penis', True)

    def test_superfluous_int_argument_raises_type_error(self):
        with self.assertRaises(TypeError):
            mkdong(10, False, False, 20)

    def test_superfluous_str_argument_raises_type_error(self):
        with self.assertRaises(TypeError):
            mkdong(10, False, False, 'cock')


mkdong/core.py
#!/usr/bin/env python

"""
Prints a dong of a specific length, width, and climax.
"""


from collections import namedtuple
import os
import sys
import argparse

from .version import __version__


__author__ = 'Jathan McCollum, Mark Ellzey Thomas, Clint Fralick'
__maintainer__ = 'Jathan McCollum'
__email__ = 'jathan@gmail.com'


# Static immutable penis parts
BALL = '( )'
HEAD = 'D'
SHAFT_THIN = '='
SHAFT_WIDE = '/'
CLIMAX = ' ~~~~'

ENV_VAR_NAME = 'MEGADONG'
NUM_BALLS = 2
DEFAULT_MAXLEN = 40

MAXLEN = os.environ.get(ENV_VAR_NAME) or DEFAULT_MAXLEN


#: Used to store dong-making information
DongTuple = namedtuple('DongTuple', 'length wide climax')


def parse_args():
    """
    Parses args and returns a ``DongTuple``.

    :returns:
        `~mkdong.DongTuple`
    """
    parser = argparse.ArgumentParser(
        description='Prints a dong.', prog='mkdong',
        epilog='mkdong %s' % __version__
    )

    parser.add_argument(
        'length',
        metavar='<length>',
        type=int,
        help='The desired dong length.'
    )
    parser.add_argument(
        '-w', '--wide',
        action='store_true',
        help='Make a wide, thick dong.'
    )
    parser.add_argument(
        '-c', '--climax',
        action='store_true',
        help='Makes the dong climax.'
    )
    parser.add_argument(
        '-v', '--version',
        action='version',
        version='%(prog)s ' + __version__,
        help='Print version and exit.'
    )
    dong_args = parser.parse_args()

    if dong_args.length > MAXLEN:
        parser.error(
            # "error: a %s\" dong is too big! cannot be longer than %s\"!" %
            'Soul brother, too beaucoup! (Cannot be longer than %s.)' % MAXLEN
        )

    return DongTuple(dong_args.length, dong_args.wide, dong_args.climax)


def mkshaft(length, wide=False):
    """
    Generate the shaft.

    :param length:
        Desired length

    :param wide:
        Wide or no?

    :returns:
        Shaft
    """
    shaft = SHAFT_WIDE if wide else SHAFT_THIN
    return _mk_dong_part(shaft, length, '')


def mkballs(balls=NUM_BALLS):
    """
    Generate the balls.

    :param balls:
        Number of balls

    :returns:
        Balls
    """
    return _mk_dong_part(BALL, balls, '/')


def _mk_dong_part(value, num_items, delimiter):
    """
    Join dong parts together.

    :param value:
        Dong part value

    :param num_items:
        Dong part repetitions

    :param delimiter:
        Dong part delimiter

    :returns:
        Dong part
    """
    return delimiter.join(value for _ in xrange(num_items))


def construct_dong(balls, shaft, head, climax=False):
    """
    Construct the dong.

    :param balls:
        Balls

    :param shaft:
        Shaft

    :param head:
        Head

    :param climax:
        Climax or no?

    :returns:
        Dong string
    """
    dong_parts = [balls, shaft, head]
    if climax:
        dong_parts.append(CLIMAX)

    return ''.join(dong_parts)


def mkdong(length, wide=False, climax=False):
    """
    Returns a fully constructed dong.

    :param length:
        Desired length

    :param wide:
        Wide or no?

    :param climax:
        Whether to climax

    :returns:
        Dong
    """
    shaft = mkshaft(length, wide)
    balls = mkballs()
    head = HEAD
    return construct_dong(balls, shaft, head, climax)


def main():
    """The Main Event."""
    length, wide, climax = parse_args()

    try:
        dong = mkdong(length=length, wide=wide, climax=climax)
    except ValueError as ronjeremy:
        sys.exit(str(ronjeremy))
    else:
        print dong


mkdong/mkdong.py
#!/usr/bin/env python

"""
Prints a dong of a specific length.
"""


import os
import sys
import argparse


__author__ = 'Jathan McCollum, Mark Ellzey Thomas, Clint Fralick'
__maintainer__ = 'Jathan McCollum'
__email__ = 'jathan@gmail.com'
__version__ = '6.1'


# Static immutable penis parts
BALL = '( )'
HEAD = 'D'
THIN = '='
WIDE = '/'
CLIMAX = ' ~~~~'

ENV_VAR_NAME = 'MEGADONG'
DEFAULT_MAXLEN = 40

MAXLEN = os.environ.get(ENV_VAR_NAME) or DEFAULT_MAXLEN


def parse_args():
    parser = argparse.ArgumentParser(
        description='Prints a dong.',
        prog='mkdong',
        epilog='mkdong %s' % __version__
    )

    parser.add_argument('length',
                        metavar='l',
                        type=int,
                        help='the desired dong length')
    parser.add_argument('-w',
                        '--wide',
                        action='store_true',
                        help='make a wide, thick dong')
    parser.add_argument('-c',
                        '--climax',
                        action='store_true',
                        help='makes the dong climax')
    parser.add_argument('-v',
                        '--version',
                        action='version',
                        version='%(prog)s ' + __version__,
                        help='print version and exit')
    dong_args = parser.parse_args()

    if dong_args.length > MAXLEN:
        parser.error(
            "error: a %s\" dong is too big! cannot be longer than %s\"!" %
            (dong_args.length, MAXLEN)
        )

    return (dong_args.length, dong_args.wide, dong_args.climax,)


def mkshaft(length, width):
    w = WIDE if width else THIN
    return _mk_dong_part(w, length, '')


def mkballs(balls=2):
    return _mk_dong_part(BALL, balls, '/')


def _mk_dong_part(v, r, d):
    return d.join([v for x in xrange(r)])


def _mk_dong(b, s, h, c=None):
    if c:
        d = ''.join([b, s, h, CLIMAX])
    else:
        d = ''.join([b, s, h])
    return d


def mkdong(length, width, climax=None):
    """Prints a dong of ``length`` length."""
    shaft = mkshaft(length, width)
    balls = mkballs()
    return _mk_dong(b=balls, s=shaft, h=HEAD, c=climax)


def main():
    length, width, climax = parse_args()

    try:
        dong = mkdong(length=length, width=width, climax=climax)
    except ValueError as ronjeremy:
        sys.exit(str(ronjeremy))
    else:
        os.system('echo "%s"' % dong)
